src/message.c

```diff
@@ -1109,95 +1146,97 @@ qd_message_t *qd_message_receive(pn_delivery_t *delivery)
+                        // pending buffer has bytes that are port of message
+                        DEQ_INSERT_TAIL(msg->content->buffers,
+                                        msg->content->pending);
+                    } else {
+                        // pending buffer is empty
+                        qd_buffer_free(msg->content->pending);
+                    }
+                    msg->content->pending = 0;
+                } else {
+                    // pending buffer is absent
+                }
+
+                msg->content->receive_complete = true;
 
-            rc = pn_link_recv(link, (char*) qd_buffer_cursor(buf), qd_buffer_capacity(buf));
+                // unlink message and delivery
+                pn_record_set(record, PN_DELIVERY_CTX, 0);
+            }
+            sys_mutex_unlock(msg->content->lock);
+            return (qd_message_t*) msg;
         }
 
         //
-        // If we receive PN_EOS, we have come to the end of the message.
+        // Handle a missing or full pending buffer
         //
-        if (rc == PN_EOS) {
-            //
-            // We have received the entire message since rc == PN_EOS, set the receive_complete flag to true
-            //
-            msg->content->receive_complete = true;
-
-            //
-            // Clear the value in the record with key PN_DELIVERY_CTX
-            //
-            pn_record_set(record, PN_DELIVERY_CTX, 0);
-
-            //
-            // If the last buffer in the list is empty, remove it and free it.  This
-            // will only happen if the size of the message content is an exact multiple
-            // of the buffer size.
-            //
-            if (buf && qd_buffer_size(buf) == 0) {
+        if (!msg->content->pending) {
+            // Pending buffer is absent: get a new one
+            msg->content->pending = qd_buffer();
+        } else {
+            // Pending buffer exists
+            if (qd_buffer_capacity(msg->content->pending) == 0) {
+                // Pending buffer is full
                 sys_mutex_lock(msg->content->lock);
-                DEQ_REMOVE_TAIL(msg->content->buffers);
+                DEQ_INSERT_TAIL(msg->content->buffers, msg->content->pending);
                 sys_mutex_unlock(msg->content->lock);
-                qd_buffer_free(buf);
+                msg->content->pending = qd_buffer();
+            } else {
+                // Pending buffer still has capacity
             }
-
-            return (qd_message_t*) msg;
         }
 
-        if (rc > 0) {
-            if (discard)
-                continue;
+        //
+        // Try to fill the remaining space in the pending buffer.
+        //
+        rc = pn_link_recv(link,
+                          (char*) qd_buffer_cursor(msg->content->pending),
+                          qd_buffer_capacity(msg->content->pending));
+
+        assert (rc != PN_EOS); // Just checked for this moments ago
```

Comment:
This assert is a mistake. Between sensing it a moment ago and now it may have become true.